### DISCUSSION TITLE
How to add GEN Charge Battery current & Gen force for DEYE ?

### DISCUSSION THREAD

**Discussion by man55** [kellerza/sunsynk#318] [Q&A]
Hi, thanks again for the great integration!

I have DEYE-SUN-6K-SG03LP1-EU WiFi

It would be nice to add a settings sensor to adjust the charging current from the generator. I already added "Grid Charge Battery current" according to the Guide, but I didn't find how to add a similar option for the generator "GEN Charge Battery current".

I would also like to add “Gen force”, this option is available in the inverter, it forces the generator to start.
Generator(AUX input) voltage and frequency would also be very helpful.

I read the Custom sensors section, but I don’t even understand how to approach this.

**Comment by man55** [comment]
I figured it out when I re-read the custom sensors section a few more times.
I was able to add the sensors I needed.
The most difficult thing in this is to understand the Chinese PDF in search of the necessary sensor numbers.

**Reply by kellerza** [reply]
The modbus document is not the most pleasant read. It could be inverter & version dependent and may also depend on you install. Do you use it as an AUX out/have a generator connected, etc.

Glad you got it working. If you have any tips for the community on having a generator connected, and suggestions to the docs would be appreciated.

**Reply by man55** [reply]
> Glad you got it working. If you have any tips for the community on having a generator connected, and suggestions to the docs would be appreciated.

I found some sensors&settings and tested their work on my DEYE-SUN-6K-SG03LP1-EU:
```
    NumberRWSensor(226, "GEN starting capacity", "%", max=50),
    Sensor(181, "Generator voltage", VOLT, 0.1),
    Sensor(195, "Generator status"),
    Sensor(196, "Generator frequency", "Hz", 0.01),
    NumberRWSensor(227, "GEN charge Battery current", AMPS, min=0, max=135),
```

These work, but have not yet been sufficiently tested in practice:
```
    NumberRWSensor(223, "GEN max operating time", "h", 0.01, min=0, max=4),  // operating time before shutdown for cooling
    NumberRWSensor(224, "GEN cooling time", "h", 0.01, min=0, max=2),  // generator cooling down time
    Sensor(62, "Day GEN Power", KWH, 0.1),
    Sensor(83, "Day GEN Time", "h", 0.1),
```

The most interesting thing is Generator status, which takes on different values ​​depending on the status: 
16  off, 
272  signaled to start, 
273  the generator is running and supplies voltage to the inverter.
This is convenient to use for creating relevant notifications.
